From 563467d8aa5b1db95ee0795f2e76909d5573bafa Mon Sep 17 00:00:00 2001
From: Peng Ma <peng.ma@nxp.com>
Date: Thu, 13 Dec 2018 18:58:57 +0800
Subject: [PATCH 406/706] dmaengine: fsl-edma: support little endian for edma
 driver

improve edma driver to support little endian.

Signed-off-by: Peng Ma <peng.ma@nxp.com>
(cherry picked from commit 6e65c7b3d59752112ccb3ab68ae943f0c16ef1b8)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/dma/fsl-edma.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/dma/fsl-edma.c b/drivers/dma/fsl-edma.c
index c7568869284e..ef0da1479496 100644
--- a/drivers/dma/fsl-edma.c
+++ b/drivers/dma/fsl-edma.c
@@ -256,9 +256,14 @@ static void fsl_edma_chan_mux(struct fsl_edma_chan *fsl_chan,
 	u32 ch = fsl_chan->vchan.chan.chan_id;
 	void __iomem *muxaddr;
 	unsigned chans_per_mux, ch_off;
+	int endian_diff[4] = {3, 1, -1, -3};
 
 	chans_per_mux = fsl_chan->edma->n_chans / DMAMUX_NR;
 	ch_off = fsl_chan->vchan.chan.chan_id % chans_per_mux;
+
+	if (!fsl_chan->edma->big_endian)
+		ch_off += endian_diff[ch_off % 4];
+
 	muxaddr = fsl_chan->edma->muxbase[ch / chans_per_mux];
 	slot = EDMAMUX_CHCFG_SOURCE(slot);
 
-- 
2.17.1

